Additions to major-rewriting

rowanthorpe opened the pull request:
These are various tweaks, fixes, suggested improvements from proofreading the major-rewriting branch for #56. The commit-messages should be self-explanatory.

GioBonvi commented:
Looks like a lot of great work!

Will look into it as soon as possible.

Thanks

GioBonvi commented on code.gs:
```diff
@@ -156,6 +168,13 @@ var settings = {
+    /* NB: Users shouldn't need to (or want to) touch these settings. They are here for the
+     *     convenience of developers/maintainers only.
+     */
+    version: '3.2.0-alpha',
```
Moving the version number from [line L845](https://github.com/GioBonvi/GoogleContactsEventsNotifier/pull/60/files#diff-deb2c49f6ef99255c9830fe80e966b46L845) to here requires [line R1173](https://github.com/GioBonvi/GoogleContactsEventsNotifier/pull/60/files#diff-deb2c49f6ef99255c9830fe80e966b46R1173) to be modified with an updated regex.

However, since this would not be the first time that moving the version number around breaks this functionality for old versions, I think we should be using [GitHub's API](https://developer.github.com/v3/repos/releases/#get-the-latest-release) instead. (In fact I don't know why I did not think of this to start with...).  
The latest version number would be obtained simply by GETting and parsing this JSON file: https://api.github.com/repos/GioBonvi/GoogleContactsEventsNotifier/releases/latest which should be both more reliable and stable across releases and would still let use move the current version number around as we like.

Let me know what you think of this: if you want and have the time you can implement this in a new commit in this PR, otherwise I'll do it once this PR is merged.

GioBonvi replied:
Moving the version number here would require [line R1173](https://github.com/GioBonvi/GoogleContactsEventsNotifier/pull/60/files#diff-deb2c49f6ef99255c9830fe80e966b46R1173) to be updated with a new RegEx.

However, since changing the RegEx would break older version update detection and this is not the first time that this happens I think we should be using [GitHub's API](https://developer.github.com/v3/repos/releases/#get-the-latest-release) to detect the latest released version of the script, since it would be more reliable (no RegEx, just JSON parsing), more stable (no changes breaking retrocompatibility) and would still let use move the _current version number_ around as we please.

Let me know if you want to add this to the PR in a new commit, otherwise I'll do it when this PR is merged.

GioBonvi commented on code.gs:
```diff
@@ -530,35 +554,53 @@ DataCollector.prototype.isPropEmpty = function (key) {
-    }).length === 0;
-};
+    }).length !== 0;
+}
```
A semicolon is required here as per SemistandardJS style.

GioBonvi replied:
SemistandardJS requires a semicolon here.

rowanthorpe commented:
@GioBonvi Thanks for catching those two issues. I fixed the coding-style typo, and added a commit for checking the script-version from the API. I tested this version using `unitTests()`, and ran the `test()` twice, using an up-to-date version-number and a deliberately out-of-date one - and got the expected result for each.

BTW: I finally just now found a "js semistandard" minor-mode for Emacs, so I should be able to catch my style-violations from now on automatically :-)

rowanthorpe commented:
@GioBonvi: Now I have the semistandard-autoformatter running in my Emacs I found a `throw(` which had became `throw (`, so just pushed that change too now. BTW: the autoformatter trims the comma from the trailing entries in all the i18n tables. I personally prefer the trailing commas so that translators don't have to worry about e.g.
```
'a': 'x',
'b': 'y'
```
accidentally becoming:
```
'a': 'x',
'b': 'y'
'c': 'z'
```
but I thought I would check with you first, before I do one of:

- fix those and push them
- work out how to configure exceptions when running the `semistandard` tool

GioBonvi commented:
Yes, I decided to keep the commas in i18n for that exact reason. I don't use a formatter, buy a linter, so I just ignore those errors.  
I don't know if there is a inter-plugin way to configure the rules (and I suspect there is not).

rowanthorpe commented:
Now I worked out how to run the semistandard linter (not the formatter) it also showed me an error with the `throw` expecting an object so I pushed `throw ('')` -> `throw new Error('')` too.
